hardhat/contracts/Fundraiser.sol

```diff
@@ -0,0 +1,385 @@
+  struct Campaign {
+    bytes32 id;
+    Account account;
+    string dataCID;
+    CampaignKind kind;
+    bytes32 categoryId;
+    uint256 startAt;
+    uint256 endAt;
+    uint256 goal;
+    uint256 donated;
+    uint256 paid;
+    uint256 numDonations;
+    CampaignState state;
+  }
+
+  struct Account {
+    bytes32 id;
+    address owner;
+    string dataCID;
+    AccountKind kind;
+    VerificationState verification;
+  }
+
+  struct Donation {
+    uint256 amount;
+    uint256 timestamp;
+    address sender;
+  }
+
+  struct CampaignCategory {
+    bytes32 id;
+    string name;
+  }
+
+  struct Stats {
+    uint256 totalAccounts;
+    uint256 totalCampaigns;
+    uint256 totalDonations;
+    uint256 totalDonated;
+  }
+
+  mapping(bytes32 => Campaign) public campaigns;
+  mapping(address => Account) public accounts;
+  mapping(bytes32 => Campaign[]) public campaignsByAccount;
+  mapping(bytes32 => string) public campaignCategories;
+  mapping(bytes32 => Donation[]) public donationsByCampaign;
+
+  address public paymentController;
+  Stats private _stats;
+  bytes32[] private _campaignCategoryIds;
+  string[] private _defaultCampaignCategories = [
+    'Environment',
+    'Animals',
+    'Emergencies',
+    'Family',
+    'Business',
+    'Events',
+    'Community',
+    'Education',
+    'Memorials',
+    'Medical',
+    'Bills',
+    'Travel',
+    'Sports',
+    'Volunteer',
+    'Other'
+  ];
+
+  modifier controllerOnly() {
+    require(_msgSender() == paymentController, "CONTROLLER_UNAUTHORIZED");
+    _;
+  }
+
+  modifier pauserOnly() {
+    require(hasRole(PAUSER_ROLE, _msgSender()), "PAUSER_UNAUTHORIZED");
+    _;
+  }
+
+  modifier adminOnly() {
+    require(hasRole(ADMIN_ROLE, _msgSender()), "ADMIN_UNAUTHORIZED");
+    _;
+  }
+
+  modifier campaignManagerOnly() {
+    Account storage account = accounts[_msgSender()];
+    require(account.owner == _msgSender() || hasRole(ADMIN_ROLE, _msgSender()), "UNAUTHORIZED");
+    _;
+  }
+
+  modifier accountOwnerOnly() {
+    Account storage account = accounts[_msgSender()];
+    require(account.owner == _msgSender(), "SENDER_UNAUTHORIZED");
+    _;
+  }
+
+  event CampaignCreated(Campaign campaign, uint256 createdAt);
+  event CampaignUpdated(Campaign campaign, uint256 updatedAt);
+  event AccountCreated(bytes32 indexed accountId, string accountDataCID, address indexed owner, uint256 timestamp);
+  event CampaignCategoryCreated(bytes32 indexed categoryId, string indexed name, address indexed caller);
+  event ControllerChanged(address previousController, address newController);
+  event DonationMade(uint256 amount, uint256 timestamp, address indexed caller);
+  event DistributedFunds(bytes32 campaignId, uint256 amount, address indexed caller, uint256 timestamp);
+  event CampaignClosed(bytes32 campaignId, address indexed caller, uint256 timestamp);
+  event CampaignPublished(bytes32 campaignId, address indexed caller, uint256 timestamp);
+
+  constructor(address _paymentController) {
```

Comment:
I am including the idea of a separate contract handling the payment distributions in the the future, which is why the constructor is being initialized with an address. For now it is not being used in order to establish a working proof of concept and MVP for hackathon deadline.

Reply:
From what I understand, when starting a Campaign, we just pick an address to where it goes, and we assume this location is a "trusted" location to dump tokens.  Adding complexity of trust validation is for future implementation.